{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.6.12 64-bit ('maxATAC_binary_CNN': conda)",
   "metadata": {
    "interpreter": {
     "hash": "249b28acec6c509c36f85a5c161ac4762b548c3e465e9dde4848a9c879dd0d05"
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from maxatac.utilities.genome_tools import RandomRegionsGenerator\n",
    "from maxatac.utilities.genome_tools import build_chrom_sizes_dict\n",
    "import types\n"
   ]
  },
  {
   "source": [
    "## Initialize the RandomRegionsGenerator\n",
    "\n",
    "The generator takes as input:\n",
    "\n",
    "<pre>\n",
    "chrom_sizes_dict,\n",
    "chrom_pool_size,\n",
    "region_length,\n",
    "sequence,\n",
    "average,\n",
    "meta_table,\n",
    "input_channels,\n",
    "bp_resolution=BP_RESOLUTION,\n",
    "method=\"length\"\n",
    "</pre>"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "args = types.SimpleNamespace()\n",
    "\n",
    "args.chrom_pool_size = 1000\n",
    "args.region_length = 1024\n",
    "args.sequence = \"/Users/caz3so/scratch/maxATAC/data/genome_inf/hg38/hg38.2bit\"\n",
    "args.average = \"/Users/caz3so/scratch/maxATAC/data/genome_inf/hg38/hg38_zeroes_average_file.bw\"\n",
    "args.meta_table = \"/Users/caz3so/scratch/maxATAC/analysis/20201228_maxATAC_interactive/20201230_maxATAC_CTCF_TEST_META.tsv\"\n",
    "args.input_channels = 6\n",
    "args.bp_resolution = 1\n",
    "args.method = \"length\"\n",
    "args.chrom_sizes = \"/Users/caz3so/scratch/maxATAC/data/genome_inf/hg38/hg38.chrom.sizes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "RR_gen = RandomRegionsGenerator(build_chrom_sizes_dict([\"chr1\"], args.chrom_sizes),\n",
    "                                chrom_pool_size=args.chrom_pool_size,\n",
    "                                region_length=args.region_length,\n",
    "                                sequence=args.sequence,\n",
    "                                average=args.average,\n",
    "                                meta_table=args.meta_table,\n",
    "                                input_channels=args.input_channels,\n",
    "                                bp_resolution=args.bp_resolution,\n",
    "                                method=args.method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "[('chr1', 248956422),\n",
       " ('chr1', 248956422),\n",
       " ('chr1', 248956422),\n",
       " ('chr1', 248956422),\n",
       " ('chr1', 248956422),\n",
       " ('chr1', 248956422),\n",
       " ('chr1', 248956422),\n",
       " ('chr1', 248956422),\n",
       " ('chr1', 248956422),\n",
       " ('chr1', 248956422)]"
      ]
     },
     "metadata": {},
     "execution_count": 13
    }
   ],
   "source": [
    "RR_gen.chrom_pool[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "  Cell_Line                                   ATAC_Signal_File    TF  \\\n",
       "0      A549  /Users/caz3so/scratch/20201127_maxATAC_refined...  CTCF   \n",
       "1   GM12878  /Users/caz3so/scratch/20201127_maxATAC_refined...  CTCF   \n",
       "2   GM23338  /Users/caz3so/scratch/20201127_maxATAC_refined...  CTCF   \n",
       "3    HEK293  /Users/caz3so/scratch/20201127_maxATAC_refined...  CTCF   \n",
       "4     HepG2  /Users/caz3so/scratch/20201127_maxATAC_refined...  CTCF   \n",
       "5    IMR-90  /Users/caz3so/scratch/20201127_maxATAC_refined...  CTCF   \n",
       "6      K562  /Users/caz3so/scratch/20201127_maxATAC_refined...  CTCF   \n",
       "7     MCF-7  /Users/caz3so/scratch/20201127_maxATAC_refined...  CTCF   \n",
       "\n",
       "                                        Binding_File  \\\n",
       "0  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "1  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "2  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "3  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "4  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "5  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "6  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "7  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "\n",
       "                                          ATAC_Peaks  \\\n",
       "0  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "1  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "2  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "3  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "4  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "5  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "6  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "7  /Users/caz3so/scratch/maxATAC/analysis/2020122...   \n",
       "\n",
       "                                          CHIP_Peaks  \n",
       "0  /Users/caz3so/scratch/maxATAC/analysis/2020122...  \n",
       "1  /Users/caz3so/scratch/maxATAC/analysis/2020122...  \n",
       "2  /Users/caz3so/scratch/maxATAC/analysis/2020122...  \n",
       "3  /Users/caz3so/scratch/maxATAC/analysis/2020122...  \n",
       "4  /Users/caz3so/scratch/maxATAC/analysis/2020122...  \n",
       "5  /Users/caz3so/scratch/maxATAC/analysis/2020122...  \n",
       "6  /Users/caz3so/scratch/maxATAC/analysis/2020122...  \n",
       "7  /Users/caz3so/scratch/maxATAC/analysis/2020122...  "
      ],
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Cell_Line</th>\n      <th>ATAC_Signal_File</th>\n      <th>TF</th>\n      <th>Binding_File</th>\n      <th>ATAC_Peaks</th>\n      <th>CHIP_Peaks</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>A549</td>\n      <td>/Users/caz3so/scratch/20201127_maxATAC_refined...</td>\n      <td>CTCF</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>GM12878</td>\n      <td>/Users/caz3so/scratch/20201127_maxATAC_refined...</td>\n      <td>CTCF</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>GM23338</td>\n      <td>/Users/caz3so/scratch/20201127_maxATAC_refined...</td>\n      <td>CTCF</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>HEK293</td>\n      <td>/Users/caz3so/scratch/20201127_maxATAC_refined...</td>\n      <td>CTCF</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>HepG2</td>\n      <td>/Users/caz3so/scratch/20201127_maxATAC_refined...</td>\n      <td>CTCF</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>IMR-90</td>\n      <td>/Users/caz3so/scratch/20201127_maxATAC_refined...</td>\n      <td>CTCF</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>K562</td>\n      <td>/Users/caz3so/scratch/20201127_maxATAC_refined...</td>\n      <td>CTCF</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>MCF-7</td>\n      <td>/Users/caz3so/scratch/20201127_maxATAC_refined...</td>\n      <td>CTCF</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n      <td>/Users/caz3so/scratch/maxATAC/analysis/2020122...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "metadata": {},
     "execution_count": 14
    }
   ],
   "source": [
    "RR_gen.MetaDF[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "('chr1', 42389921, 42390945)"
      ]
     },
     "metadata": {},
     "execution_count": 15
    }
   ],
   "source": [
    "RR_gen.get_region()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "(array([[[1., 0., 0., 0., 0., 0.],\n",
       "         [1., 0., 0., 0., 0., 0.],\n",
       "         [1., 0., 0., 0., 0., 0.],\n",
       "         ...,\n",
       "         [0., 0., 1., 0., 0., 0.],\n",
       "         [1., 0., 0., 0., 0., 0.],\n",
       "         [0., 0., 0., 1., 0., 0.]]]),\n",
       " array([[0., 0., 0., ..., 0., 0., 0.]]))"
      ]
     },
     "metadata": {},
     "execution_count": 17
    }
   ],
   "source": [
    "next(RR_gen.BatchGenerator(n_rand=1))"
   ]
  },
  {
   "source": [
    "class RandomRegionsGenerator(object):\n",
    "    \"\"\"\n",
    "    This class will generate a pool of random regions\n",
    "\n",
    "    The RandomRegionsGenerator will generate a random region of interest based\n",
    "    based on the input reference genome chrom sizes, the desired size of the\n",
    "    chromosome pool, the input length, and the method used to calculate the\n",
    "    frequency of chromosome examples.\n",
    "\n",
    "    Args\n",
    "    ----\n",
    "    chrom_sizes_dict (dict):\n",
    "        A dictionary of chromosome sizes.\n",
    "    chrom_pool_size (int):\n",
    "        The size of the chromosome pool to output.\n",
    "    region_length (int):\n",
    "        The size of the regions to generate.\n",
    "    sequence (str):\n",
    "        The 2bit DNA sequence path.\n",
    "    average (str):\n",
    "        The average ATAC-seq sequence path.\n",
    "    meta_table (str):\n",
    "        Path to the meta run file.\n",
    "    input_channels (int):\n",
    "        The number of input channels.\n",
    "    bp_resolution (int):\n",
    "        The resolution of the data to use.\n",
    "    method (str): optional\n",
    "        Method for calculating chromosome frequencies.\n",
    "\n",
    "    Attributes\n",
    "    ----------\n",
    "    chrom_sizes_dict (dict):\n",
    "        A dictionary of chromosome sizes filtered for chromosomes of interest.\n",
    "    chrom_pool_size (int):\n",
    "        The size of the chromosome pool to output.\n",
    "    region_length (int):\n",
    "        The size of the regions to generate.\n",
    "    method (str):\n",
    "        Method for calculating chromosome frequencies.\n",
    "    chrom_pool (array):\n",
    "        Chromosome pool that is adjusted by frequency.\n",
    "\n",
    "    Methods\n",
    "    -------\n",
    "    get_region()\n",
    "        Gets a random region from the available pool\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(\n",
    "            self,\n",
    "            chrom_sizes_dict,\n",
    "            chrom_pool_size,\n",
    "            region_length,\n",
    "            sequence,\n",
    "            average,\n",
    "            meta_table,\n",
    "            input_channels,\n",
    "            bp_resolution=BP_RESOLUTION,\n",
    "            method=\"length\"\n",
    "    ):\n",
    "        self.chrom_sizes_dict = chrom_sizes_dict\n",
    "        self.chrom_pool_size = chrom_pool_size\n",
    "        self.region_length = region_length\n",
    "        self.__idx = 0\n",
    "        self.method = method\n",
    "        self.sequence = sequence\n",
    "        self.average = average\n",
    "        self.MetaPath = meta_table\n",
    "        self.input_channels = input_channels\n",
    "        self.bp_resolution = bp_resolution\n",
    "\n",
    "        self.chrom_pool = self.__get_chrom_frequencies()\n",
    "        self.MetaDF = self.ImportMeta()\n",
    "        self.CellTypes = self.get_CellTypes()\n",
    "        self.TranscriptionFactor = self.get_TranscriptionFactor()\n",
    "\n",
    "    def ImportMeta(self):\n",
    "        return pd.read_csv(self.MetaPath, sep='\\t', header=0, index_col=None)\n",
    "\n",
    "    def get_CellTypes(self):\n",
    "        return self.MetaDF[\"Cell_Line\"].unique().tolist()\n",
    "\n",
    "    def get_TranscriptionFactor(self):\n",
    "        return self.MetaDF[\"TF\"].unique()[0]\n",
    "\n",
    "    def get_input_matrix(self,\n",
    "                         rows,\n",
    "                         cols,\n",
    "                         batch_size,  # make sure that cols % batch_size == 0\n",
    "                         signal_stream,\n",
    "                         average_stream,\n",
    "                         sequence_stream,\n",
    "                         bp_order,\n",
    "                         chrom,\n",
    "                         start,  # end - start = cols\n",
    "                         end,\n",
    "                         reshape=True,\n",
    "\n",
    "                         ):\n",
    "\n",
    "        input_matrix = np.zeros((rows, cols))\n",
    "\n",
    "        for n, bp in enumerate(bp_order):\n",
    "            input_matrix[n, :] = get_one_hot_encoded(\n",
    "                sequence_stream.sequence(chrom, start, end),\n",
    "                bp\n",
    "            )\n",
    "\n",
    "        signal_array = np.array(signal_stream.values(chrom, start, end))\n",
    "        avg_array = np.array(average_stream.values(chrom, start, end))\n",
    "        input_matrix[4, :] = signal_array\n",
    "        input_matrix[5, :] = input_matrix[4, :] - avg_array\n",
    "        input_matrix = input_matrix.T\n",
    "\n",
    "        if reshape:\n",
    "            input_matrix = np.reshape(\n",
    "                input_matrix,\n",
    "                (batch_size, round(cols / batch_size), rows)\n",
    "            )\n",
    "\n",
    "        return input_matrix\n",
    "\n",
    "    def __get_chrom_frequencies(self):\n",
    "        \"\"\"\n",
    "        Generate an array of chromosome frequencies\n",
    "\n",
    "        The frequencies will be used to build the pool of regions that\n",
    "        we pull random regions from. The frequencies are determined by\n",
    "        the method attribute. The methods are length or proportion.\n",
    "\n",
    "        The length method will generate the frequencies of examples in\n",
    "        the pool based on the length of the chromosomes in total.\n",
    "\n",
    "        The proportionmethod will generate a pool that has chromosome\n",
    "        frequencies equel to the chromosome pools size divided by the\n",
    "        number of chromosomes.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        labels (list):\n",
    "            A list of chromosome names the size of the desired pool.\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "        if self.method == \"length\":\n",
    "            sum_lengths = sum(self.chrom_sizes_dict.values())\n",
    "\n",
    "            frequencies = {\n",
    "                chrom_name: round(chrom_len / sum_lengths * self.chrom_pool_size)\n",
    "                for chrom_name, chrom_len in self.chrom_sizes_dict.items()\n",
    "            }\n",
    "\n",
    "        else:\n",
    "            chrom_number = len(self.chrom_sizes_dict.values())\n",
    "\n",
    "            frequencies = {\n",
    "                chrom_name: round(self.chrom_pool_size / chrom_number)\n",
    "                for chrom_name, chrom_len in self.chrom_sizes_dict.items()\n",
    "            }\n",
    "\n",
    "        labels = []\n",
    "\n",
    "        for k, v in frequencies.items():\n",
    "            labels += [(k, self.chrom_sizes_dict[k])] * v\n",
    "        random.shuffle(labels)\n",
    "\n",
    "        return labels\n",
    "\n",
    "    def get_region(self):\n",
    "        \"\"\"\n",
    "        Get a random region from the chromosome pool.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        chrom_name (str):\n",
    "            The chromosome string\n",
    "        start (int):\n",
    "            The region start\n",
    "        end (int):\n",
    "            The region end\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "        # If the __idx reaches the pool size before enough samples are selected\n",
    "        # shuffle and set to 0\n",
    "        if self.__idx == self.chrom_pool_size:\n",
    "            random.shuffle(self.chrom_pool)\n",
    "            self.__idx = 0\n",
    "\n",
    "        chrom_name, chrom_length = self.chrom_pool[self.__idx]\n",
    "        self.__idx += 1\n",
    "\n",
    "        start = round(\n",
    "            random.randint(\n",
    "                0,\n",
    "                chrom_length - self.region_length\n",
    "            )\n",
    "        )\n",
    "\n",
    "        end = start + self.region_length\n",
    "\n",
    "        return chrom_name, start, end\n",
    "\n",
    "    def BatchGenerator(self, n_rand):\n",
    "        while True:\n",
    "            inputs_batch, targets_batch = [], []\n",
    "            for idx in range(n_rand):\n",
    "                cell_line = random.choice(self.CellTypes)  # Randomly select a cell line\n",
    "\n",
    "                chrom_name, start, end = self.get_region()  # returns random region (chrom_name, start, end)\n",
    "\n",
    "                meta_row = self.MetaDF[self.MetaDF['Cell_Line'] == cell_line]\n",
    "\n",
    "                meta_row = meta_row.reset_index(drop=True)\n",
    "\n",
    "                signal = meta_row.loc[0, 'ATAC_Signal_File']\n",
    "\n",
    "                binding = meta_row.loc[0, 'Binding_File']\n",
    "\n",
    "                with \\\n",
    "                        load_bigwig(self.average) as average_stream, \\\n",
    "                        load_2bit(self.sequence) as sequence_stream, \\\n",
    "                        load_bigwig(signal) as signal_stream, \\\n",
    "                        load_bigwig(binding) as binding_stream:\n",
    "\n",
    "                    input_matrix = self.get_input_matrix(\n",
    "                        rows=self.input_channels,\n",
    "                        cols=self.region_length,\n",
    "                        batch_size=1,  # we will combine into batch later\n",
    "                        reshape=False,\n",
    "                        bp_order=[\"A\", \"C\", \"G\", \"T\"],\n",
    "                        signal_stream=signal_stream,\n",
    "                        average_stream=average_stream,\n",
    "                        sequence_stream=sequence_stream,\n",
    "                        chrom=chrom_name,\n",
    "                        start=start,\n",
    "                        end=end\n",
    "                    )\n",
    "                    inputs_batch.append(input_matrix)\n",
    "                    target_vector = np.array(binding_stream.values(chrom_name, start, end)).T\n",
    "                    target_vector = np.nan_to_num(target_vector, 0.0)\n",
    "                    n_bins = int(target_vector.shape[0] / self.bp_resolution)\n",
    "                    split_targets = np.array(np.split(target_vector, n_bins, axis=0))\n",
    "                    bin_sums = np.sum(split_targets, axis=1)\n",
    "                    bin_vector = np.where(bin_sums > 0.5 * self.bp_resolution, 1.0, 0.0)\n",
    "                    targets_batch.append(bin_vector)\n",
    "\n",
    "            yield (np.array(inputs_batch), np.array(targets_batch))\n"
   ],
   "cell_type": "code",
   "metadata": {},
   "execution_count": null,
   "outputs": []
  }
 ]
}